tsfc/coffee_mode.py

```diff
@@ -121,36 +163,37 @@ def find_optimal_atomics(monomials, argument_indices):
     """
     atomic_index = defaultdict(partial(next, itertools.count()))  # atomic -> int
     connections = []
-    # add connections (list of tuples, items in each tuple form a product)
+    # add connections (list of sets, items in each tuple form a product)
     for monomial in monomials:
-        connections.append(tuple(map(lambda a: atomic_index[a], monomial.atomics)))
+        connections.append(set(map(lambda a: atomic_index[a], monomial.atomics)))
 
-    if len(atomic_index) <= 1:
+    num_atomics = len(atomic_index)
+    if num_atomics <= 1:
         return tuple(iterkeys(atomic_index))
 
-    # set up the ILP
-    import pulp as ilp
-    ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
-    ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
-
-    # Objective function
-    # Minimise number of factors to pull. If same number, favour factor with larger extent
-    penalty = 2 * max(index_extent(atomic, argument_indices) for atomic in atomic_index) * len(atomic_index)
-    ilp_prob += ilp.lpSum(ilp_var[index] * (penalty - index_extent(atomic, argument_indices))
-                          for atomic, index in iteritems(atomic_index))
-
-    # constraints
-    for connection in connections:
-        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
-
-    ilp_prob.solve()
-    if ilp_prob.status != 1:
-        raise RuntimeError("Something bad happened during ILP")
-
-    def optimal(atomic):
-        return ilp_var[atomic_index[atomic]].value() == 1
+    def calculate_extent(solution):
+        extent = 0
+        for atomic, index in iteritems(atomic_index):
+            if index in solution:
+                extent += index_extent(atomic, argument_indices)
+        return extent
```

Comment:
I note that this code seems unnecessarily complicated.  In this sense:

Previously, when you were passing data off to an external IP solver, you had to set up the problem with data structures it understood (i.e. integer variables, etc...).  But now you control all the data, so there's no need to mimic the data structures that pulp required.  In particular, `solution` is a set of integers that are mapped (via `atomic_index`) to the atomic factors we're trying to pull apart.  Would it not be much simpler if `solution` was just the set of current proposed atomics instead?

Then this could becomes:

```python
def calculate_extent(solution):
    return sum(map(lambda atomic: index_extent(atomic, argument_indices), solution))
```
?